src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +319,99 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeoutDelay q = do
+  timedOutVar ← registerDelay timeoutDelay
+
+  let
+    readItems xs =
+      readTBMQueue q ≫= \case
+        Nothing → return xs
+        Just x → go (x:xs)
+
+    timeout =
+      readTVar timedOutVar ≫= check
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = readItems xs <|> xs <$ timeout
+
+  atomically $ go []
```

Comment:
Yeah, I think that will be necessary. It also then necessitates our (hypothesized) queue structure to live in STM rather than IO.

Sent from my iPad

> On Mar 27, 2015, at 5:32 PM, Lars Kuhtz <notifications@github.com> wrote:
> 
> In src/Aws/Kinesis/Client/Producer.hs:
> 
> > +  timedOutVar ← registerDelay timeoutDelay
> > +
> > +  let
> > +    readItems xs =
> > +      readTBMQueue q ≫= \case
> > +        Nothing → return xs
> > +        Just x → go (x:xs)
> > +
> > +    timeout =
> > +      readTVar timedOutVar ≫= check
> > +
> > +    go xs
> > +      | length xs ≥ n = return xs
> > +      | otherwise = readItems xs <|> xs <$ timeout
> > +
> > +  atomically $ go []
> You are right that's an issue. A solution could be include the timeout check into the atomically block. This is a pretty cheap check that shouldn't affect the odds of go to win the race on the queue too much.
> 
> —
> Reply to this email directly or view it on GitHub.
>